src/gz/bufread.rs

```diff
@@ -40,6 +40,90 @@ fn read_le_u16<R: Read>(r: &mut R) -> io::Result<u16> {
+                r.read_once(&mut header)?;
+                r.part.buf.truncate(0);
+
+                if header[0] != 0x1f || header[1] != 0x8b {
+                    return Err(bad_header());
+                }
+                if header[2] != 8 {
+                    return Err(bad_header());
+                }
+
+                r.part.flg = header[3];
+                r.part.mtime = ((header[4] as u32) << 0)
+                    | ((header[5] as u32) << 8)
+                    | ((header[6] as u32) << 16)
+                    | ((header[7] as u32) << 24);
+                let _xfl = header[8];
+                r.part.os = header[9];
+                r.part.state = GzHeaderParsingState::GzHeaderParsingXlen;
+            }
+            GzHeaderParsingState::GzHeaderParsingXlen => {
+                if r.part.flg & FEXTRA != 0 {
+                    r.part.xlen = read_le_u16_once(r)?;
+                }
+                r.part.state = GzHeaderParsingState::GzHeaderParsingExtra;
+            }
+            GzHeaderParsingState::GzHeaderParsingExtra => {
+                if r.part.flg & FEXTRA != 0 {
+                    let mut extra = vec![0; r.part.xlen as usize];
+                    r.read_once(&mut extra)?;
+                    r.part.extra = Some(extra);
+                }
+                r.part.state = GzHeaderParsingState::GzHeaderParsingFilename;
+            }
+            GzHeaderParsingState::GzHeaderParsingFilename => {
+                if r.part.flg & FNAME != 0 {
+                    for byte in r.reader.bytes() {
+                        let byte = byte?;
+                        r.crc.update(&[byte]);
+                        if byte == 0 {
+                            break;
+                        }
+                        r.part.filename.push(byte);
+                    }
+                }
+                r.part.state = GzHeaderParsingState::GzHeaderParsingComment;
+            }
+            GzHeaderParsingState::GzHeaderParsingComment => {
+                if r.part.flg & FCOMMENT != 0 {
+                    for byte in r.reader.bytes() {
+                        let byte = byte?;
+                        r.crc.update(&[byte]);
+                        if byte == 0 {
+                            break;
+                        }
+                        r.part.comment.push(byte);
+                    }
+                }
+                r.part.state = GzHeaderParsingState::GzHeaderParsingCrc;
+            }
+            GzHeaderParsingState::GzHeaderParsingCrc => {
+                if r.part.flg & FHCRC != 0 {
+                    let stored_crc = read_le_u16_once(r)?;
+                    let calced_crc = r.crc.sum() as u16;
+                    if stored_crc != calced_crc {
+                        return Err(corrupt());
+                    }
+                }
+                return Ok(());
+            }
+        }
+    }
+}
+
 pub(crate) fn read_gz_header<R: Read>(r: &mut R) -> io::Result<GzHeader> {
```

Comment:
Shouldn't this function be replaced with the above one?

Reply:
Indeed.
But it does not work for now as gzip writer uses `read_gz_header` with a different kind of reader
And I do not know why writer needs to call `read_gz_header`
cf https://github.com/rust-lang/flate2-rs/pull/278#discussion_r672140753

Reply:
I don't know what that other call is doing, but I don't think there should be two functions to parse the header here.

Reply:
Indeed,
Now `read_gz_header` just calls once `read_gz_header_part